src/main/java/com/squareup/kotlinpoet/LambdaTypeName.kt

```diff
@@ -0,0 +1,81 @@
+/*
+ * Copyright (C) 2017 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.squareup.kotlinpoet
+
+class LambdaTypeName internal constructor(
+    private val receiver: TypeName? = null,
+    private val parameters: List<TypeName> = emptyList(),
+    private val returnType: TypeName = TypeName.get(Unit::class),
+    private val initializer: CodeBlock? = null,
+    nullable: Boolean = false,
+    annotations: List<AnnotationSpec> = emptyList()
+) : TypeName(nullable, annotations) {
+  override fun asNullable() = LambdaTypeName(receiver, parameters, returnType, initializer, true, annotations)
+
+  override fun asNonNullable() = LambdaTypeName(receiver, parameters, returnType, initializer, false, annotations)
+
+  override fun annotated(annotations: List<AnnotationSpec>): TypeName
+      = LambdaTypeName(receiver, parameters, returnType, initializer, nullable, annotations)
+
+  override fun withoutAnnotations(): TypeName = LambdaTypeName(receiver, parameters, returnType, initializer, nullable)
+
+  override fun abstractEmit(out: CodeWriter): CodeWriter {
+    annotations.forEach { it.emit(out, true) }
+    if (nullable) {
+      out.emit("(")
+    }
+
+    receiver?.let {
+      out.emit("%T.", it)
+    }
+
+    out.emit("(")
+    if (parameters.isNotEmpty()) {
```

Comment:
nit: this `if` block has no real value. drop it. (sorry should have noticed earlier)